### PR TITLE
Update LLM_HF example

### PR DESCRIPTION
**Pull request by ZiyueXu77**
Fixes # .

### Description

Update LLM_HF example to use the latest Llama-3.2 model, update to comply with recent HF API changes.
Replace job template with JobAPI script.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

**Comment by holgerroth** [comment]
Let's have a separate PR for LLM_HF update and compression filter.

**Comment by ZiyueXu77** [comment]
> Let's have a separate PR for LLM_HF update and compression filter.

moved compression part out for the next PR

File: examples/advanced/llm_hf/README.md [hunk 1]
```diff
@@ -99,55 +83,37 @@ The major code modifications are for receiving and returning the global model (r
...
-nvflare job list_templates
-```
-We can see the "sag_pt" template is available, with which we further generate job configs for SFT and PEFT as:
-```
-nvflare job create -force -j "./jobs/hf_peft" -w "sag_pt" -sd "code" \
-  -f meta.conf min_clients=1 \
-  -f config_fed_client.conf app_script="hf_sft_peft_fl.py" app_config="--model_path ${PWD}/model/Llama-2-7b-hf --data_path_train ${PWD}/dataset/dolly/training.jsonl --data_path_valid ${PWD}/dataset/dolly/validation.jsonl --output_path llama2-7b-dolly-peft --mode 1" \
-  -f config_fed_server.conf model_class_path="hf_peft_model.CausalLMPEFTModel" components[0].args.model.args.model_path="${PWD}/model/Llama-2-7b-hf" min_clients=1 num_rounds=3 key_metric="eval_loss" negate_key_metric=True
+### Federated Training Results
+We run the federated training on a single client using NVFlare Simulator via [JobAPI](../job_api/README.md).
```
Comment (holgerroth) [inline examples/advanced/llm_hf/README.md:87]: better link to the docs https://nvflare.readthedocs.io/en/main/programming_guide/fed_job_api.html

File: examples/advanced/llm_hf/README.md [hunk 2]
```diff
@@ -4,15 +4,15 @@ This example shows how to use [NVIDIA FLARE](https://nvidia.github.io/NVFlare) f
...
+We used the [Llama-3.2-1B model](https://huggingface.co/meta-llama/Llama-3.2-1B) to showcase the functionality of federated SFT and PEFT, allowing HuggingFace models to be trained and adapted with NVFlare. All other models from HuggingFace can be easily adapted following the same steps.
 
 For PEFT, we used LoRA method, other PEFT methods (e.g. p-tuning, prompt-tuning) can be easily adapted as well by modifying the configs following [PEFT](https://github.com/huggingface/peft) examples.
 
-Mainly on two fronts:
-- Adapt local HuggingFace training scripts to federated application
-- Handling large model weights (~26 GB for Llama-2-7b-hf model), this is supported by NVFlare infrastructure, and does not need any code change.
+We would like to showcase three key points in this example:
+- Adapt local HuggingFace training scripts, both SFT and PEFT, to federated application
+- Handling large model weights (~6 GB for Llama-3.2-1B model with float32 precision for communication), which is beyond protobuf's 2 GB hard limit. It is supported by NVFlare infrastructure via streaming, and does not need any code change.
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/advanced/llm_hf/README.md:13]: yes, float32 is 6 GB message

File: examples/advanced/llm_hf/README.md [hunk 3]
```diff
@@ -4,15 +4,15 @@ This example shows how to use [NVIDIA FLARE](https://nvidia.github.io/NVFlare) f
...
 
-We used the [Llama-2-7b-hf model](https://huggingface.co/meta-llama/Llama-2-7b-hf) to showcase the functionality of federated SFT and PEFT, allowing HuggingFace models to be trained and adapted with NVFlare. 
+We used the [Llama-3.2-1B model](https://huggingface.co/meta-llama/Llama-3.2-1B) to showcase the functionality of federated SFT and PEFT, allowing HuggingFace models to be trained and adapted with NVFlare. All other models from HuggingFace can be easily adapted following the same steps.
 
 For PEFT, we used LoRA method, other PEFT methods (e.g. p-tuning, prompt-tuning) can be easily adapted as well by modifying the configs following [PEFT](https://github.com/huggingface/peft) examples.
 
-Mainly on two fronts:
-- Adapt local HuggingFace training scripts to federated application
-- Handling large model weights (~26 GB for Llama-2-7b-hf model), this is supported by NVFlare infrastructure, and does not need any code change.
+We would like to showcase three key points in this example:
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/advanced/llm_hf/README.md:11]: two, updated

File: examples/advanced/llm_hf/README.md (hunk 1 above)
Comment (ZiyueXu77) [reply examples/advanced/llm_hf/README.md:87]: thanks, updated

File: examples/advanced/llm_hf/sft_job.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        "--data_path",
+        type=str,
+        default="",
+        help="root directory for training and validation data",
+    )
+    parser.add_argument(
+        "--train_mode",
+        type=int,
+        default=0,
+        help="training mode, 0: SFT, 1: PEFT, default to 0",
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/advanced/llm_hf/sft_job.py:138]: makes sense, changed

File: examples/advanced/llm_hf/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (holgerroth) [reply examples/advanced/llm_hf/README.md:13]: ok. Why use float32 at all here? to_numpy conversion shouldn't change the data type.

[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/advanced/llm_hf/README.md:13]: Training is done in BF16, which has a similar dynamic range to FP32, while FP16 has a narrower range. So to best keep the range fidelity in a simple way, it needs to be casted to FP32.

[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/advanced/llm_hf/README.md:13]: For BF16,

```
app_opt/pt/params_converter.py", line 51, in convert
    return_tensors[k] = v.cpu().numpy()
TypeError: Got unsupported ScalarType BFloat16
```

[the remark this answers is not in the document]
Comment (holgerroth) [reply examples/advanced/llm_hf/README.md:13]: Why not send bf16 tensors directly then? We should be able to aggregate them, right?

[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/advanced/llm_hf/README.md:13]: No, not before we can send tensor directly

**Review by holgerroth — APPROVED** [review]
Good job. Thanks!

**Pull request merged**